src/core/stdc/stdarg.d

```diff
@@ -218,261 +193,6 @@ else version (X86_64)
         return a;
     }
 
-    ///
-    void va_arg(T)(va_list apx, ref T parmn)
-    {
-        __va_list* ap = cast(__va_list*) apx;
```

Comment:
Could this instead be:
```
///
void va_arg(T)(va_list apx, ref T parmn)
{
    __va_arg_internal!T(apx, parmn);
}
```
With main body of implementation moved as appropriate?

Reply:
This is marked as WIP because it's not finished yet, there's a 2nd restructuring upcoming, just waiting for this first commit to turn green.

Reply:
It's never too early to get suggestions in. :-)

Reply:
The first commit wasn't going to get green due to something mitigated in the 2nd, so pushed earlier. - I suppose GDC has a bunch of `va_arg` extensions similar to LDC; can you give me a link so that I can cross-check and try to upstream at least some more platforms?

Reply:
It's all using built-ins, there's a version(GNU) block with all the signatures, but no bodies.

Reply:
Oh wow, that's convenient.

Reply:
> Oh wow, that's convenient.

- `__builtin_va_list` is constructed during language init from a gcc tree node made available by backend init (it is then inserted into the gcc.builtins module)
- va_copy and va_end are aliasing `__builtin_` symbols from gcc.builtins.
- va_start and va_arg are the only specially handled intrinsics, the calls to the template instantiations are matched and replaced with `__builtin_va_start` and `VA_ARG_EXPR` respectively.

The target takes care of providing the actual implementations, meaning that va_args just works on every platform and I do sod all in order to make that happen. :-)

The only omission would be the rtti based va_arg function, which relies on either the language parts doing something target specific (bad abstraction/separation of concerns), or implementing a language specific backend hook that duplicates/fleshes out the internal type -> mode inference done when backend lowers va_arg (duplication is a code smell at best, and fleshing out internals needs approval from each maintainer for each backend  that we need this, which may hit upon strong resistance).

When I spoke to the C frontend (and global gcc) maintainer about this, his immediate reaction was that I/we should be using libffi, and if it doesn't fit our case, then we can fix it so that it can be used for our purposes.

Reply:
LLVM doesn't make it that easy, unfortunately.

- I think we can safely agree that the TypeInfo-based `va_arg` should be deprecated; there was a single failure in the whole testsuite as long as that was missing for LDC and AArch64 (and that was supposed to test something completely different IIRC), so it's pretty much untested anyway.
- As for the regular `va_arg(T)` template, I think DMD's `argTypes` solution for complex vararg implementations isn't bad, acting as a bridge between compiler and druntime (it's not a trait, but the special `is(T ArgTypes == __argTypes)` expression), in order to have a more legible implementation of `va_arg` in druntime rather than having to emit the equivalent in each compiler via intrinsic. Unless of course the intrinsic/builtin is already handled as in your lucky case :), then I wouldn't use the argTypes at all either.

Reply:
> LLVM doesn't make it that easy, unfortunately.
> 
>  **edit**: yeah well and D-style variadic functions

There's nothing inherently wrong with D style variadics, type inference using only RTTI is just a pain if you want to support structs.

Reply:
I just imagine them to be a pain without according TypeInfo-based `va_arg`, even if just using primitive types. I might have written some of these variadics at some point for a test, but never had any real need for them.